added logging for slow RPC requests

dvovk opened the pull request:
Changed distribution of httpcfg.HttpCfg to be pointer.
Added new flags:
rpc.slow.log - which is false by default, this flag need to enable logging slow RPC requests
rpc.slow.log.threshold - which is 100 by default, this flag specify slow threshold in milliseconds 
Updated rpc handler to log slow requests:
- added map[request id] {method, timestamp}
- put every request details to map above
- delete request details from map above
- added time interval check for elements in map and if time difference is more than given threshold print request id and the method
- app will print slow requests in next cases:
1. As soon as request take more than given threshold
2. Every 20 seconds if request still in process
3. After request finished and it took more than give threshold

mh0lt reviewed (commented):
I don't think you need both flags, you can just add the threshold and use  threshold!=0 as the flag.

mh0lt reviewed (commented):
I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?

dvovk commented:
> I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?
> 
> I think it makes the map logic a bit more complex. I think you should keep a start and end time and then remove it from the map once its logged.

I think frequency of log should depend on threshold value as all sense of this logs is to print slow requests which isn't finished yet. I agree that I can change logic to print slow request only on the first check

dvovk commented on rpc/handler.go:
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+	}
+
+	quit := make(chan struct{})
+	go func() {
```
if I do it it just hangs in the loop inside this coroutine and never process request

AskAlexSharov replied:
Because you didn’t listen rootctx.Done()

dvovk commented:
> I don't think you need both flags, you can just add the threshold and use threshold!=0 as the flag.

Fixed

dvovk commented on rpc/handler.go:
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+					}
+				}
+			case <-quit:
+				ticker.Stop()
```
Fixed

dvovk commented on rpc/handler.go:
```diff
@@ -128,15 +139,49 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+		updateThreshold = slowThreshold
+	}
+
+	quit := make(chan struct{})
```
Fixed

dvovk commented:
> > I don't think we should run this log message 10 times per second - it will be pretty expensive I'd have thought we want to run it say once every 20 secs ?
> > I think it makes the map logic a bit more complex. I think you should keep a start and end time and then remove it from the map once its logged.
> 
> I think frequency of log should depend on threshold value as all sense of this logs is to print slow requests which isn't finished yet. I agree that I can change logic to print slow request only on the first check

Fixed, added 20 second log interval

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+
+		rpcSlowLogThreshold: rpcSlowLogThreshold,
+		diagMessage:         diagMsg{Method: "", Time: time.Unix(0, 0), LastLogTime: time.Unix(0, 0), Finished: true},
+		rpcMethodsWhitelist: []string{"eth_call", "eth_getBlock"},
```
blacklist

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+			select {
+			case <-ticker.C:
+				h.diagLock.Lock()
+				if h.isRpcMethodNeedsCheck(h.diagMessage.Method) {
```
`isRpcMethodNeedsCheck` check it once - before starting goroutine and ticker.

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+	}
+
+	go func() {
+		ticker := time.NewTicker(updateThreshold)
```
why not `logInterval`?

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+					if !h.diagMessage.Finished {
+						requestDuration := time.Since(h.diagMessage.Time)
+						if requestDuration > slowThreshold && time.Since(h.diagMessage.LastLogTime) > logInterval {
+							h.logger.Warn("[Slow RPC] - request still processing", "method", h.diagMessage.Method, "Request duration", requestDuration)
```
let's use prefix `[rpc.slow]` - then it will be more grep-friendly

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -70,13 +70,26 @@ type handler struct {
+
+	// metrics
+	diagMessage         diagMsg
+	diagLock            sync.Mutex
```
looks like we don't need this lock anymore

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -128,15 +142,64 @@ func newHandler(connCtx context.Context, conn jsonWriter, idgen func() ID, reg *
+func (h *handler) lookForSlowRpc() {
+	logInterval := 30 * time.Second
+	updateThreshold := 100 * time.Millisecond
+	slowThreshold := time.Duration(h.rpcSlowLogThreshold) * time.Millisecond
```
this can be done once - when parsing CLI flag.

dvovk commented:
@AskAlexSharov  I rethink it a bit and removed a lot of stuff, I think now it is much better

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
 		return nil
 	case msg.isCall():
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
+			slowTimer := time.AfterFunc(time.Duration(h.rpcSlowLogThreshold)*time.Millisecond, func() {
```
really interesting. looks good for me

let's calc `time.Duration(h.rpcSlowLogThreshold)*time.Millisecond` once - when parsing cli flag.

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
+
 		resp := h.handleCall(ctx, msg, stream)
+
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
```
let's store result of `h.isRpcMethodNeedsCheck(msg.Method)` into local bool variable

AskAlexSharov commented on rpc/handler.go:
```diff
@@ -388,7 +406,23 @@ func (h *handler) handleCallMsg(ctx *callProc, msg *jsonrpcMessage, stream *json
 	case msg.isCall():
+		if h.rpcSlowLogThreshold > 0 && h.isRpcMethodNeedsCheck(msg.Method) {
+			slowTimer := time.AfterFunc(time.Duration(h.rpcSlowLogThreshold)*time.Millisecond, func() {
+				h.logger.Warn("[rpc.slow] - running", "method", msg.Method)
```
is it possible to print request ID and request body here? (and print only ID at `finish`)

dvovk commented:
@AskAlexSharov changed mentioned stuff

AskAlexSharov commented:
did push some changes - plz review

dvovk commented:
> did push some changes - plz review

Looks ok
